File: 魏晋南北朝系列/scripts/generate-markdown.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
生成 100 篇 Markdown 文章的 Python 脚本
运行：python scripts/generate-markdown.py
"""

import os
import re
import json

# 读取 articles-data.js 中的文章数据
DATA_DIR = os.path.join(os.path.dirname(__file__), '..', 'site', 'js')
DATA_FILE = os.path.join(DATA_DIR, 'articles-data.js')

with open(DATA_FILE, 'r', encoding='utf-8') as f:
    content = f.read()

# 提取 CATEGORIES 数据
categories_match = re.search(r'const CATEGORIES = (\[.*?\]);', content, re.DOTALL)
if categories_match:
    categories_str = categories_match.group(1)
    # 将 JavaScript 对象转换为 Python 兼容格式
    categories_str = categories_str.replace("'", '"')
    categories_str = re.sub(r'(\w+):\s*', r'"\1":', categories_str)
    # 使用 json 模块解析
    # 但 js 对象有 id/name 属性，需要处理
    categories = []
    for item_match in re.finditer(r"\{[^}]+\}", categories_str):
        item_str = item_match.group()
        id_match = re.search(r'"id"\s*:\s*"([^"]+)"', item_str)
        name_match = re.search(r'"name"\s*:\s*"([^"]+)"', item_str)
        if id_match and name_match:
            categories.append((id_match.group(1), name_match.group(1)))

# 提取 ARTICLES 数据
# 将 JavaScript 对象数组转换为 Python 列表
articles_raw = re.search(r'const ARTICLES = (\[.*?\]);\s*$', content, re.DOTALL)
if not articles_raw:
    # 找到 ARTICLES 定义结束位置
    start = content.find('const ARTICLES = [')
    if start >= 0:
        # 找到最后一个 ] 前面
        end = content.rfind('];') + 1
        articles_str = content[start:end]
    else:
        articles_str = ''
else:
    articles_str = articles_raw.group(1)

# 提取每个文章对象
articles = []

# 更简单的处理方式：直接读取 HTML 文件中的内容
# 实际上我们直接从 js 文件中提取每个 article 的字段
article_pattern = re.compile(
    r"\{\s*id:\s*(\d+),\s*category:\s*'([^']+)',\s*eras:\s*\[([^\]]+)\],\s*title:\s*'([^']*)',\s*summary:\s*'([^']*)',\s*content:\s*`([^`]+)`,\s*order:\s*(\d+)\s*\}"
)

for match in article_pattern.finditer(content):
    article_id = int(match.group(1))
    category = match.group(2)
    eras_str = match.group(3)
    title = match.group(4)
    summary = match.group(5)
    body = match.group(6)
    
    eras = [e.strip().strip("'") for e in eras_str.split(',')]
    
    articles.append({
        'id': article_id,
        'category': category,
        'eras': eras,
        'title': title,
        'summary': summary,
        'content': body
    })

# 按照 id 排序
articles.sort(key=lambda x: x['id'])

# 获取分类名称
def get_cat_name(cat_id):
    for c_id, c_name in categories:
        if c_id == cat_id:
            return c_name
    return cat_id

# 对内容进行格式化：将 HTML 标签转为 Markdown
def html_to_markdown(html):
    text = html
    text = text.replace('<h3>', '### ')
    text = text.replace('</h3>', '\n')
    text = text.replace('<strong>', '**')
    text = text.replace('</strong>', '**')
    text = text.replace('<p>', '')
    text = text.replace('</p>', '\n\n')
    return text

# 生成目录
def generate_toc():
    md = '# 魏晋南北朝系列 · 三百年风云\n\n'
    md += '## 📜 文章目录\n\n'
    md += '从曹魏到隋朝统一 · 公元220年–589年 · 共100篇\n\n'
    md += '---\n\n'
    
    for cat_id, cat_name in categories:
        cat_articles = [a for a in articles if a['category'] == cat_id]
        if not cat_articles:
            continue
        md += f'### {cat_name}\n\n'
        for a in cat_articles:
            filename = f"{a['id']:03d}-{re.sub(r'[：:，。、？\s/]+', '-', a['title']).strip('-')}.md"
            md += f'- **第{a["id"]}篇** [{a["title"]}]({filename})  \n  {a["summary"]}\n\n'
        md += '---\n\n'
    
    return md

# 生成单篇文章
def generate_article(article):
    cat_name = get_cat_name(article['category'])
    body = html_to_markdown(article['content'])
    
    md = f'# 第{article["id"]}篇 · {article["title"]}\n\n'
    md += f'> **专题**：{cat_name}  \n'
    md += f'> **标签**：{"、".join(article["eras"])}\n\n'
    md += '---\n\n'
    md += body.strip()
    md += '\n\n---\n\n'
    md += f'*本文为"魏晋南北朝系列"第{article["id"]}篇。*\n'
    
    return md

# 写入文件
articles_dir = os.path.join(os.path.dirname(__file__), '..', 'articles')
os.makedirs(articles_dir, exist_ok=True)

# 生成目录
toc = generate_toc()
toc_path = os.path.join(articles_dir, 'README.md')
with open(toc_path, 'w', encoding='utf-8') as f:
    f.write(toc)
print(f'✅ 目录已生成: articles/README.md')

# 生成每一篇文章
for article in articles:
    safe_title = re.sub(r'[：:，。、？\s/]+', '-', article['title']).strip('-').rstrip('-')
    filename = f"{article['id']:03d}-{safe_title}.md"
    filepath = os.path.join(articles_dir, filename)
    
    content = generate_article(article)
    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(content)
    print(f'✅ 第{article["id"]:3d}篇已生成: articles/{filename}')

print(f'\n🎉 全部 {len(articles)} 篇 Markdown 文章已生成完毕！')

File: 魏晋南北朝系列/scripts/generate-markdown.js
// ===== 生成 Markdown 源文件 =====
// 在 Node.js 环境下运行：node generate-markdown.js
// 将生成 100 篇 Markdown 文章到 articles/ 目录

const fs = require('fs');
const path = require('path');

// 复制 articles-data.js 中的数据结构（简化版）
const CATEGORIES = [
    { id: 'intro', name: '总论' },
    { id: 'caowei', name: '曹魏' },
    { id: 'shuhan-dongwu', name: '蜀汉·东吴' },
    { id: 'xijin', name: '西晋' },
    { id: 'dongjin', name: '东晋' },
    { id: 'shiliuguo', name: '十六国' },
    { id: 'nanchao-song', name: '南朝·宋' },
    { id: 'nanchao-qiliangchen', name: '南朝·齐梁陈' },
    { id: 'beichao-wei', name: '北朝·北魏' },
    { id: 'beichao-dongxibei', name: '北朝·东西魏北齐周' },
    { id: 'wenhua', name: '文化思想' },
    { id: 'zhidu', name: '制度经济' },
    { id: 'suichao', name: '隋朝统一' }
];

// 读取文章数据
const dataPath = path.join(__dirname, '..', 'site', 'js', 'articles-data.js');
const dataContent = fs.readFileSync(dataPath, 'utf-8');

// 用 eval 提取数据（简单方式）
// 由于文件中定义了 const，我们使用 require 的方式
// 实际上我们需要把文件临时改成一个模块，或者用 Function eval
// 简单起见直接从字符串中正则提取 eval
const evalCode = dataContent
    .replace('const CATEGORIES', 'var CATEGORIES')
    .replace('const TIMELINE', 'var TIMELINE')
    .replace('const DYNASTIES', 'var DYNASTIES')
    .replace('const ARTICLES', 'var ARTICLES');

eval(evalCode);

// 生成目录
function generateTOC() {
    let md = `# 魏晋南北朝系列 · 三百年风云\n\n`;
    md += `## 📜 文章目录\n\n`;
    md += `从曹魏到隋朝统一 · 公元220年–589年 · 共100篇\n\n`;
    md += `---\n\n`;

    for (const cat of CATEGORIES) {
        const articles = ARTICLES.filter(a => a.category === cat.id);
        if (articles.length === 0) continue;

        md += `### ${cat.name}\n\n`;
        for (const article of articles) {
            md += `- **第${article.id}篇** [${article.title}](${String(article.id).padStart(3, '0')}-${article.title.replace(/[：:，。、？\s]+/g, '-')}.md)  \n  ${article.summary}\n\n`;
        }
        md += `---\n\n`;
    }

    return md;
}

// 生成单篇文章
function generateArticle(article) {
    const cat = CATEGORIES.find(c => c.id === article.category);
    const categoryName = cat ? cat.name : article.eras[0];

    // 将 HTML 标签转换回 Markdown 格式
    let body = article.content
        .replace(/<h3>/g, '### ')
        .replace(/<\/h3>/g, '\n')
        .replace(/<strong>/g, '**')
        .replace(/<\/strong>/g, '**')
        .replace(/<p>/g, '')
        .replace(/<\/p>/g, '\n\n');

    let md = `# 第${article.id}篇 · ${article.title}\n\n`;
    md += `> **专题**：${categoryName}  \n`;
    md += `> **标签**：${article.eras.join('、')}\n\n`;
    md += `---\n\n`;
    md += body;
    md += `\n\n---\n\n`;
    md += `*本文为"魏晋南北朝系列"第${article.id}篇。*\n`;

    return md;
}

// 写入文件
const articlesDir = path.join(__dirname, '..', 'articles');
if (!fs.existsSync(articlesDir)) {
    fs.mkdirSync(articlesDir, { recursive: true });
}

// 生成目录
const toc = generateTOC();
fs.writeFileSync(path.join(articlesDir, 'README.md'), toc, 'utf-8');
console.log('✅ 目录已生成: articles/README.md');

// 生成每一篇文章
for (const article of ARTICLES) {
    const filename = `${String(article.id).padStart(3, '0')}-${article.title.replace(/[：:，。、？\s]+/g, '-')}.md`;
    const content = generateArticle(article);
    fs.writeFileSync(path.join(articlesDir, filename), content, 'utf-8');
    console.log(`✅ 第${article.id}篇已生成: articles/${filename}`);
}

console.log('\n🎉 全部100篇Markdown文章已生成完毕！');